File: android/app/src/main/kotlin/com/bansheerun/BansheeAudioManager.kt
package com.bansheerun

import android.content.Context
import android.media.AudioAttributes
import android.media.AudioFocusRequest
import android.media.AudioManager
import android.media.MediaPlayer
import android.media.SoundPool
import android.os.Build
import android.os.Handler
import android.os.Looper

/**
 * Manages scary audio effects synchronized with visual banshee weather overlay.
 * Plays ambient music and triggered sound effects when the runner falls behind.
 */
class BansheeAudioManager(private val context: Context) {

    private val audioManager = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
    private var mediaPlayer: MediaPlayer? = null
    private var soundPool: SoundPool? = null
    private val handler = Handler(Looper.getMainLooper())

    // Sound IDs for SoundPool
    private var bansheeWailId = 0
    private var whispersId = 0
    private var heartbeatId = 0

    // Current state
    private var currentIntensity = 0f
    private var isPlaying = false
    private var hasAudioFocus = false

    // Cooldowns for sound effects (prevent spam)
    private var lastWailTime = 0L
    private var lastWhisperTime = 0L
    private var lastHeartbeatTime = 0L

    // Stream IDs for stopping sounds
    private var heartbeatStreamId = 0

    private val focusRequest: AudioFocusRequest by lazy {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            AudioFocusRequest.Builder(AudioManager.AUDIOFOCUS_GAIN_TRANSIENT_MAY_DUCK)
                .setAudioAttributes(
                    AudioAttributes.Builder()
                        .setUsage(AudioAttributes.USAGE_GAME)
                        .setContentType(AudioAttributes.CONTENT_TYPE_SONIFICATION)
                        .build()
                )
                .setOnAudioFocusChangeListener { focusChange ->
                    when (focusChange) {
                        AudioManager.AUDIOFOCUS_LOSS,
                        AudioManager.AUDIOFOCUS_LOSS_TRANSIENT -> {
                            hasAudioFocus = false
                            pauseAudio()
                        }
                        AudioManager.AUDIOFOCUS_GAIN -> {
                            hasAudioFocus = true
                            if (isPlaying) resumeAudio()
                        }
                    }
                }
                .build()
        } else {
            throw IllegalStateException("AudioFocusRequest requires API 26+")
        }
    }

    init {
        initializeSoundPool()
    }

    private fun initializeSoundPool() {
        val attributes = AudioAttributes.Builder()
            .setUsage(AudioAttributes.USAGE_GAME)
            .setContentType(AudioAttributes.CONTENT_TYPE_SONIFICATION)
            .build()

        soundPool = SoundPool.Builder()
            .setMaxStreams(4)
            .setAudioAttributes(attributes)
            .build()

        soundPool?.let { pool ->
            bansheeWailId = pool.load(context, R.raw.banshee_wail, 1)
            whispersId = pool.load(context, R.raw.whispers, 1)
            heartbeatId = pool.load(context, R.raw.heartbeat, 1)
        }
    }

    /**
     * Set the audio intensity (0.0 to 1.0) synchronized with visual effects.
     * This controls volume and triggers sound effects based on intensity level.
     */
    fun setIntensity(intensity: Float) {
        val clampedIntensity = intensity.coerceIn(0f, 1f)

        if (clampedIntensity > 0f && !isPlaying) {
            startAmbientMusic()
        } else if (clampedIntensity <= 0f && isPlaying) {
            stopScaryAudio()
            return
        }

        currentIntensity = clampedIntensity

        // Adjust ambient music volume
        mediaPlayer?.setVolume(clampedIntensity * 0.7f, clampedIntensity * 0.7f)

        // Trigger sound effects based on intensity thresholds
        val currentTime = System.currentTimeMillis()

        // Whispers at medium intensity (0.4+), cooldown 8 seconds
        if (clampedIntensity >= 0.4f && currentTime - lastWhisperTime > 8000) {
            playWhispers()
            lastWhisperTime = currentTime
        }

        // Heartbeat at higher intensity (0.6+), cooldown 5 seconds
        if (clampedIntensity >= 0.6f && currentTime - lastHeartbeatTime > 5000) {
            playHeartbeat()
            lastHeartbeatTime = currentTime
        }
    }

    /**
     * Play the banshee wail sound effect (for when significantly behind).
     * Has a 10-second cooldown to prevent spam.
     */
    fun playBansheeWail() {
        val currentTime = System.currentTimeMillis()
        if (currentTime - lastWailTime < 10000) return // 10 second cooldown

        lastWailTime = currentTime
        soundPool?.play(
            bansheeWailId,
            currentIntensity.coerceAtLeast(0.5f),
            currentIntensity.coerceAtLeast(0.5f),
            1,
            0,
            1.0f
        )
    }

    private fun playWhispers() {
        soundPool?.play(
            whispersId,
            currentIntensity * 0.6f,
            currentIntensity * 0.6f,
            1,
            0,
            1.0f
        )
    }

    private fun playHeartbeat() {
        // Stop previous heartbeat if still playing
        if (heartbeatStreamId != 0) {
            soundPool?.stop(heartbeatStreamId)
        }

        heartbeatStreamId = soundPool?.play(
            heartbeatId,
            currentIntensity * 0.8f,
            currentIntensity * 0.8f,
            1,
            0,  // no loop
            1.0f + (currentIntensity * 0.3f)  // speed up slightly with intensity
        ) ?: 0
    }

    private fun startAmbientMusic() {
        if (!requestAudioFocus()) return

        if (mediaPlayer == null) {
            mediaPlayer = MediaPlayer.create(context, R.raw.ambient_scary)?.apply {
                isLooping = true
                setVolume(0f, 0f)
            }
        }

        mediaPlayer?.let { player ->
            if (!player.isPlaying) {
                player.start()
                isPlaying = true
                // Fade in
                fadeVolume(0f, currentIntensity * 0.7f, 1000)
            }
        }
    }

    /**
     * Stop all scary audio with a fade out effect.
     */
    fun stopScaryAudio() {
        if (!isPlaying) return

        // Fade out then stop
        fadeVolume(currentIntensity * 0.7f, 0f, 500) {
            mediaPlayer?.pause()
            mediaPlayer?.seekTo(0)
            isPlaying = false
            currentIntensity = 0f
            abandonAudioFocus()

            // Stop any playing heartbeat
            if (heartbeatStreamId != 0) {
                soundPool?.stop(heartbeatStreamId)
                heartbeatStreamId = 0
            }
        }
    }

    private fun pauseAudio() {
        mediaPlayer?.pause()
    }

    private fun resumeAudio() {
        mediaPlayer?.start()
    }

    private fun fadeVolume(from: Float, to: Float, durationMs: Long, onComplete: (() -> Unit)? = null) {
        val steps = 20
        val stepDuration = durationMs / steps
        val volumeStep = (to - from) / steps

        var currentStep = 0
        var currentVolume = from

        val runnable = object : Runnable {
            override fun run() {
                if (currentStep < steps) {
                    currentVolume += volumeStep
                    mediaPlayer?.setVolume(currentVolume, currentVolume)
                    currentStep++
                    handler.postDelayed(this, stepDuration)
                } else {
                    mediaPlayer?.setVolume(to, to)
                    onComplete?.invoke()
                }
            }
        }
        handler.post(runnable)
    }

    private fun requestAudioFocus(): Boolean {
        val result = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            audioManager.requestAudioFocus(focusRequest)
        } else {
            @Suppress("DEPRECATION")
            audioManager.requestAudioFocus(
                { /* legacy callback */ },
                AudioManager.STREAM_MUSIC,
                AudioManager.AUDIOFOCUS_GAIN_TRANSIENT_MAY_DUCK
            )
        }
        hasAudioFocus = result == AudioManager.AUDIOFOCUS_REQUEST_GRANTED
        return hasAudioFocus
    }

    private fun abandonAudioFocus() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            audioManager.abandonAudioFocusRequest(focusRequest)
        } else {
            @Suppress("DEPRECATION")
            audioManager.abandonAudioFocus { /* legacy callback */ }
        }
        hasAudioFocus = false
    }

    /**
     * Release all resources. Call this when the activity/service is destroyed.
     */
    fun release() {
        stopScaryAudio()
        handler.removeCallbacksAndMessages(null)
        mediaPlayer?.release()
        mediaPlayer = null
        soundPool?.release()
        soundPool = null
    }
}

File: macos/BansheeRun/BansheeAudioManager.swift
import AVFoundation
import Foundation

/// Manages scary audio effects synchronized with visual banshee weather overlay.
/// Plays ambient music and triggered sound effects when the runner falls behind.
class BansheeAudioManager: ObservableObject {

    private var ambientPlayer: AVAudioPlayer?
    private var bansheeWailPlayer: AVAudioPlayer?
    private var whispersPlayer: AVAudioPlayer?
    private var heartbeatPlayer: AVAudioPlayer?

    @Published private(set) var isPlaying = false
    private var currentIntensity: Float = 0

    // Cooldowns for sound effects (prevent spam)
    private var lastWailTime: Date = .distantPast
    private var lastWhisperTime: Date = .distantPast
    private var lastHeartbeatTime: Date = .distantPast

    private var fadeTimer: Timer?

    init() {
        setupAudioSession()
        loadSounds()
    }

    private func setupAudioSession() {
        #if os(iOS)
        do {
            let session = AVAudioSession.sharedInstance()
            try session.setCategory(.playback, mode: .default, options: [.mixWithOthers, .duckOthers])
            try session.setActive(true)
        } catch {
            print("Failed to setup audio session: \(error)")
        }
        #endif
        // macOS doesn't need explicit audio session setup
    }

    private func loadSounds() {
        // Load ambient scary music
        if let url = Bundle.main.url(forResource: "ambient_scary", withExtension: "mp3") {
            ambientPlayer = try? AVAudioPlayer(contentsOf: url)
            ambientPlayer?.numberOfLoops = -1  // Loop indefinitely
            ambientPlayer?.prepareToPlay()
        }

        // Load banshee wail
        if let url = Bundle.main.url(forResource: "banshee_wail", withExtension: "mp3") {
            bansheeWailPlayer = try? AVAudioPlayer(contentsOf: url)
            bansheeWailPlayer?.prepareToPlay()
        }

        // Load whispers
        if let url = Bundle.main.url(forResource: "whispers", withExtension: "mp3") {
            whispersPlayer = try? AVAudioPlayer(contentsOf: url)
            whispersPlayer?.prepareToPlay()
        }

        // Load heartbeat
        if let url = Bundle.main.url(forResource: "heartbeat", withExtension: "mp3") {
            heartbeatPlayer = try? AVAudioPlayer(contentsOf: url)
            heartbeatPlayer?.prepareToPlay()
        }
    }

    /// Set the audio intensity (0.0 to 1.0) synchronized with visual effects.
    /// This controls volume and triggers sound effects based on intensity level.
    func setIntensity(_ intensity: Float) {
        let clampedIntensity = min(max(intensity, 0), 1)

        if clampedIntensity > 0 && !isPlaying {
            startAmbientMusic()
        } else if clampedIntensity <= 0 && isPlaying {
            stopScaryAudio()
            return
        }

        currentIntensity = clampedIntensity

        // Adjust ambient music volume
        ambientPlayer?.volume = clampedIntensity * 0.7

        // Trigger sound effects based on intensity thresholds
        let now = Date()

        // Whispers at medium intensity (0.4+), cooldown 8 seconds
        if clampedIntensity >= 0.4 && now.timeIntervalSince(lastWhisperTime) > 8 {
            playWhispers()
            lastWhisperTime = now
        }

        // Heartbeat at higher intensity (0.6+), cooldown 5 seconds
        if clampedIntensity >= 0.6 && now.timeIntervalSince(lastHeartbeatTime) > 5 {
            playHeartbeat()
            lastHeartbeatTime = now
        }
    }

    /// Play the banshee wail sound effect (for when significantly behind).
    /// Has a 10-second cooldown to prevent spam.
    func playBansheeWail() {
        let now = Date()
        guard now.timeIntervalSince(lastWailTime) >= 10 else { return }

        lastWailTime = now
        bansheeWailPlayer?.volume = max(currentIntensity, 0.5)
        bansheeWailPlayer?.currentTime = 0
        bansheeWailPlayer?.play()
    }

    private func playWhispers() {
        whispersPlayer?.volume = currentIntensity * 0.6
        whispersPlayer?.currentTime = 0
        whispersPlayer?.play()
    }

    private func playHeartbeat() {
        heartbeatPlayer?.volume = currentIntensity * 0.8
        heartbeatPlayer?.enableRate = true
        heartbeatPlayer?.rate = 1.0 + (currentIntensity * 0.3)
        heartbeatPlayer?.currentTime = 0
        heartbeatPlayer?.play()
    }

    private func startAmbientMusic() {
        guard let player = ambientPlayer else { return }

        player.volume = 0
        player.play()
        isPlaying = true

        // Fade in
        fadeVolume(from: 0, to: currentIntensity * 0.7, duration: 1.0)
    }

    /// Stop all scary audio with a fade out effect.
    func stopScaryAudio() {
        guard isPlaying else { return }

        fadeVolume(from: ambientPlayer?.volume ?? 0, to: 0, duration: 0.5) { [weak self] in
            self?.ambientPlayer?.pause()
            self?.ambientPlayer?.currentTime = 0
            self?.isPlaying = false
            self?.currentIntensity = 0

            // Stop other sounds
            self?.heartbeatPlayer?.stop()
            self?.whispersPlayer?.stop()
            self?.bansheeWailPlayer?.stop()
        }
    }

    private func fadeVolume(from: Float, to: Float, duration: TimeInterval, completion: (() -> Void)? = nil) {
        fadeTimer?.invalidate()

        let steps = 20
        let stepDuration = duration / Double(steps)
        let volumeStep = (to - from) / Float(steps)

        var currentStep = 0
        var currentVolume = from

        fadeTimer = Timer.scheduledTimer(withTimeInterval: stepDuration, repeats: true) { [weak self] timer in
            currentStep += 1
            currentVolume += volumeStep
            self?.ambientPlayer?.volume = currentVolume

            if currentStep >= steps {
                timer.invalidate()
                self?.ambientPlayer?.volume = to
                completion?()
            }
        }
    }

    deinit {
        fadeTimer?.invalidate()
        ambientPlayer?.stop()
        bansheeWailPlayer?.stop()
        whispersPlayer?.stop()
        heartbeatPlayer?.stop()
    }
}
